### ISSUE TITLE
Parameter collection behaviour

### ISSUE THREAD

**Issue by mdelhoneux** [clab/dynet#1265]
Not sure if this is a bug or if I'm doing something wrong but I'm having some trouble understanding the behaviour of the parameter collection object.
I have a network object with a parameter collection (self.model) which I pass to 2 MLPs. The MLPs are used for scoring during training. They are defined as their own object.
If I make the MLPs their own parameter collection, my network trains fine but otherwise it does not.
It seems to create the parameters all right (the saved model file looks fine) but the loss and errors don't go down. I guess it means that the MLP parameters don't get backpropped into.
The MLP parameters used to be defined directly inside my Network object, but I wanted to move them out into their own file to make the code easier to work with.
I can go back to that ugly version but I won't if I can avoid :)
I can make the MLPs their own parameter collection but it will be a bit more of a pain to do model saving and loading.

Any idea what might be going on?

This is a skeleton of what my code looks like:

```python
import dynet as dy

class Network(object):
    def __init__(self):
        self.model = dy.ParameterCollection()
        self.trainer = dy.AdamTrainer(self.model)
        self.mlp1 = MLP(self.model, in_dim, hid_dim, out_dim1)
        self.mlp2 = MLP(self.model, in_dim, hid_dim, out_dim2)

    def train():
        for example in examples:
            score1 = self.mlp1(example)
            score2 = self.mlp2(example)
            loss = loss_function(score1,score2)
        self.trainer.update()


class MLP(object):
    def __init__(self, model, in_dim, hid_dim, out_dim, activation=dy.tanh):
        #model = dy.ParameterCollection() # this works
        self._W1 = model.add_parameters((hid_dim, in_dim))
        self._b1 = model.add_parameters(hid_dim)
        self._W2 = model.add_parameters((out_dim, hid_dim))
        self._b2 = model.add_parameters(out_dim)
        self.activation = activation

    def hid_layer(self,x):
        W = dy.parameter(self._W1)
        b = dy.parameter(self._b1)
        return self.activation(W*x+b)

    def out_layer(self,x):
        W = dy.parameter(self._W2)
        b = dy.parameter(self._b2)
        return self.activation(W*x+b)

    def __call__(self,x):
        h = self.hid_layer(x)
        return self.out_layer(h)
```

**Comment by pmichel31415** [comment]
Make sure that your trainer gets passed the correct parameter collection (seems like it gets self.model so it will only update the parameters in self.model)

**Comment by mdelhoneux** [comment]
But self.model *is* the only parameter collection (or at least, that's what I'm trying to achieve)

**Comment by pmichel31415** [comment]
Oh OK, my bad I misread the code. It's weird though I think that your code should work and that it should break if you uncomment the `#model = dy.ParameterCollection() # this works` line.

If you can provide a bit more code (like a toy dataset) so that I can reproduce the bug I can take a look.

**Comment by mdelhoneux** [comment]
Ok. My code is part of quite a big project so I'll try to see if I can reproduce this on a toy example.

**Comment by mdelhoneux** [comment]
Update which makes even less sense, uncommenting my line ```python #model = dy.ParameterCollection() # this works``` indeed does not work but if I create a parameter collection as an attribute of the MLP it does work.

Right now I can't think of a minimal example that I could easily try out but I have published a minimal working example of the code I'm working on here: 32aa408c0ea57a22d25898bd65c245e285f8b1d6

In src/multilayer_perceptron.py I have the lines that do not work, and I commented out the lines that do work.

I call the MLP objects in src/arc_hybrid.py (can be found with #@@MLP comments line just above).
The examples can be run with test.sh
My log of the setup that does not work is in logs/setup1.log, and the one that does work is in logs/setup2.log

**Comment by pmichel31415** [comment]
OK I think I found the problem.

I printed the difference between the initial `_W1` and the current one. In the case where you put `self.model = dy.ParameterCollection() ` (the "working" case) the parameters are NOT updated. The error decreases because there must be some other parameters that are trained in your model.

In the other case, the parameters ARE updated but the error is not decreasing, my guess is that there might be a problem with the model (as in the neural model not the dynet model) itself.

So afaics the behaviour from dynet is consistent with what we would expect.

**Comment by mdelhoneux** [comment]
Ok, thanks! Did you check _W1 of both MLPs? I find it hard to imagine how the error could decrease if both MLPs don't get updated but if it's just one, I see how it could work.

**Comment by mdelhoneux** [comment]
I checked myself and observed the same you did for both MLPs. Strange that it seems to learn something at all without the MLP being updated but you're right that indeed something else must be going on in the not working version.

**Comment by mdelhoneux** [comment]
Actually that's not true, when I save the models, I get two lines of numbers per parameter, the first is the same, the second is not:
I'm not sure how this model writing works, does this matter?
See this vimdiff of the model of one of the MLPs after seeing 100 sentences vs after seeing 3000 sentences 
![parameter_files](https://user-images.githubusercontent.com/8643413/36843800-5515cd66-1d50-11e8-92d6-c15185b443f2.png)

**Comment by pmichel31415** [comment]
I think that this is normal. The second line for each parameter is the latest gradient. In this case the problem is not that the gradient is not computed but that it's not used to update the parameters

**Comment by mdelhoneux** [comment]
Looks like you're right, I also printed self.labeledMLP._W1.expr().npvalues() through learning and they don't get updated. It seems completely remarkable that the parser would learn something with 2 randomly initialised MLPs whose parameters are fixed (I trained a full epoch on a dataset I usually use and it seems to get the accuracy I would expect). I'll look more into this.

**Comment by mdelhoneux** [comment]
Got some more insight into my various issues, I'm closing this issue. Thanks for your time!
